lib/sycamore/sycamore/connectors/opensearch/utils.py

```diff
@@ -0,0 +1,21 @@
+from typing import Optional
+
+from sycamore import Context
+from sycamore.context import context_params
+from sycamore.transforms import Embedder
+
+
+@context_params("opensearch")
+def get_knn_query(text_embedder: Embedder, query_phrase: str, k: int = 500, context: Optional[Context] = None):
+    embedding = text_embedder.generate_text_embedding(query_phrase)
+    query = {
+        "query": {
+            "knn": {
```

Comment:
I've been using the "neural" feature:
https://github.com/aryn-ai/sycamore/blob/01f42f124b972211474882d0bf3392f134e799b3/apps/query-demo/querydemo/Query_Demo.py#L447

How is this difference? Should we settle on one way to do this?

Reply:
Neural search expects you to have an embedding model setup in your opensearch cluster. So the input to a neural search is a string, which opensearch will then create a vector of using that model, and then perform knn.

Knn doesn't use opensearch to create the query embedding, but instead accepts a vector as an input. So to use this we need to embed the query string outside of the opensearch call.

Knn is simpler for us since it requires less configuration on the opensearch instance, and access to the embedding model can be tricky to do in opensearch. From what I see, most vector dbs expect the latter model, so this approach unifies the logic in that sense as well.

I'd prefer knn if we think the UI will have enough juice to run an embedding model (almost certainly yes at this point)